### PR TITLE
Redis reconnect cleanup

### PR DESCRIPTION
**Pull request by mvdan**
In preparation for #520.

### CODE REVIEW

**Comment by mvdan** [comment]
Also note that this removes tail calls, which can be nothing but trouble.

**Comment by mvdan** [comment]
Also a nice coverage bump as this repeated code isn't being tested (I can't even make it run in a test).

File: redis_cluster_handler.go [hunk 1]
```diff
@@ -128,13 +128,25 @@ func (r *RedisClusterStorageManager) cleanKey(keyName string) string {
...
-func (r *RedisClusterStorageManager) GetKey(keyName string) (string, error) {
-	if GetRelevantClusterReference(r.IsCache) == nil {
-		log.Info("Connection dropped, connecting..")
+func (r *RedisClusterStorageManager) ensureConnection() {
+	if GetRelevantClusterReference(r.IsCache) != nil {
+		// already connected
+		return
+	}
+	log.Info("Connection dropped, reconnecting...")
+	for {
```
Comment (buger) [inline redis_cluster_handler.go]: Why is it inside `for` loop?
`r.Connect` can never return `false` and `GetRelevantClusterReference` seems like never can be `nil` if `Connect` was called.

Comment (mvdan) [inline redis_cluster_handler.go]: These two things are what I was talking about on slack and in the original issue.

This PR is not about changing or fixing the logic, it's about deduplicating it. I simply rewrote the recursive call into a for loop. Which makes the broken logic more obvious too.

**Comment by buger** [comment]
I see, so just to summarize:

All the storage handlers share the same Redis pool and use singletons. 
Once any of `RedisClusterStorageManager` instances makes the first `Connect()` call, it initializes singletons (and actually never returns `false`).
There are no places in code where singletons can be set to `nil`, which means all the reconnect logic does nothing and can be safely removed. To be precise, reconnecting still handled but inside `redigocluster` and `redigo`.

If we talking about #520 bug itself, I do not see the way how our code can cause them, so it is probably inside `redigocluster`. <ake sense to check if we need to update our fork.

@mvdan anything missing?

**Comment by buger** [comment]
Worth noticing that I tried simulating Redis failure, like stopping it after it was initialized, single and multiple nodes: `GetKey` always returned proper error, without looping.

**Comment by mvdan** [comment]
> All the storage handlers share the same Redis pool and use singletons.

I haven't looked at this deeply, but you pointed me towards the reconnect logic - and that only touches the singletons.

> There are no places in code where singletons can be set to nil, which means all the reconnect logic does nothing and can be safely removed. To be precise, reconnecting still handled but inside redigocluster and redigo.

Yes.

> If we talking about #520 bug itself, I do not see the way how our code can cause them, so it is probably inside redigocluster. Make sense to check if we need to update our fork.

Could be somewhere else in our redis-handling code, but likely in redigocluster, yes. Especially since we're behind on upstream changes and @lonelycode made changes that noone else is using besides tyk. So it's likely that it has obscure bugs like this one.

> Worth noticing that I tried simulating Redis failure, like stopping it after it was initialized, single and multiple nodes: GetKey always returned proper error, without looping.

I also failed to reproduce the looping. It could be that the looping is due to a race condition, though. Have you tried running your tests with `-race`? Although it's likely that it will panic due to a data race before it's able to reproduce anything.

**Comment by mvdan** [comment]
@buger is there anything else I should do or comment on?

**Comment by mvdan** [comment]
reping @buger this is for 2.3.3

**Comment by buger** [comment]
If we are talking in the context of 2.3.3, it seems like you did not change any functionality, just refactoring. I do not see how it can fix Redis connection count, or I'm missing something?

**Comment by mvdan** [comment]
Note that we haven't found the cause of the reconnect stuff, so there can't really be a fix for it.

@lonelycode mentioned getting rid of the tail calls for 2.3.3, which is what this PR is doing. It's a good first step until we can reproduce the bug.

**Pull request merged**